lib/templates/src/components/InteractiveWelcome.vue

```diff
@@ -0,0 +1,34 @@
+<template>
+  <div>
+    This screen sets up the socket connection for the interactive experiment.
+    Previously, the participant would need to agree to the terms of the
+    interactive experiment before continuing to the lobby. However, if the terms
+    are already agreed to in the previous welcome screen, you may want to
+    advance this screen automatically.
+  </div>
+</template>
+
+<script>
+import { initializeExperiment } from '@/socket.js';
+
+export default {
+  name: 'InteractiveWelcome',
+  inject: {
+    $exp: {
+      from: 'experiment',
+      default: () => ({})
+    }
+  },
+  watch: {
+    '$store.state.interactiveExperiment.experimentAvailable': function (value) {
+      if (value === true) {
+        $exp.nextScreen();
```

Comment:
@marcelklehr Could you provide an example of how to add a custom component into a generated app?

When I need to go to the next screen, apparently I need to invoke the `nextScreen()` function from `$exp`. However, I'm not sure how to inject it properly into a newly written custom component. So far, all components in the template project come from `magpie-base` itself. However, it is totally imaginable that the expeirment designer wants to write their own component, even for a non-interactive experiment. It would be nice to have a guide and ideally an actual sample component included among the templates to illustrate how to do it as well.

Reply:
I guess another point is that it seems that you usually write `$exp.nextScreen()` separately from a component itself, which makes sense in static experiments, however, here in the interactive experiment, we'd actually want the next screen to be triggered automatically, as soon as certain conditions are met. Of course, we may also do so in a more roundabout way (e.g. enable the button to be clickable as soon as certain conditions are met... Also not sure how easy it would be since the button lives outside of the component), but being able to go to the next screen automatically is ideal.

In general I'm not a fan of the `inject` approach since it's indirect and adds OOP-like dependency. If we could expose a function that any component could clearly import and use, it may seem more idiomatic (same for `addResults`? Though I understand that maybe the special structure of an experiment with an indeterminate number of screens makes that a bit harder and you probably had some rationale when choosing to do it via `inject`.

Reply:
In the demo project I just wrote `inject: ['nextScreen']` and it worked, and it was how it was like previously. Though now I see the usage generally being `$exp.nextScreen` so I wonder if the previous approach wouldn't work anymore.

Reply:
> It would be nice to have a guide and ideally an actual sample component included among the templates to illustrate how to do it as well.

that's a good point, I'll add docs for that.

Reply:
In this repo I changed the magpie boilerplate from using inject to injecting magpie automatically as a plugin, similar to how vuex or vue-router is injected in a vue app. So all components in the app automatically have `this.$exp` (or just `$exp` in the template) available. Using the inject method doesn't work anymore, because I realized that implementers would have to inject every single API, which I deemed impractical.

I did some research on whether there's a less "magical" way  to do this, but I haven't turned up anything substantial. We could refer to the Experiment component using $parent in Screen components instead and expose its API via slot parameters `<Screen #="{nextScreen, addResult}">`, but it may not always be the direct parent and it feels less intuitive to me, because I like the tangibility of having a single object that I can call on.

My reasoning was that magpie provides some global experiment state and functionality that should be available to all components (namely, "nextScreen", "addResult", "getResults", "getMouseTrack", "startMouseTracking", "submit", "trial" etc.). Similar to vue-router or vuex I found the best way to expose these APIs is via a plugin injecting the API as a special object into all components. (I'm open to other names, of course, e.g. "$magpie" or "$experiment" perhaps).

I'm aware that this is not the normal thing to do in a Vue app, but I think, given that we're building a framework for other people to write the actual apps with, this seems ok.

Does that make sense?

As to whether to use $exp directly in any custom components or expose an event is completely up to you I'd say. $exp is always available, but you can just as easily emit an event.

Reply:
My idea was that we could expose it as some module function or even similar to Vuex commits, which we would then explicitly import in each component and call. I guess I prefer the functional programming tradition of explicit over implicit. For example, without an example, I found it hard to figure out how to inject the `$exp` into a custom component, while if `nextScreen` were an explicit function/event it would always be clear. However, if with adequate documentation and example it's clear to the user how to use the `$exp` object, I am fine with it since as you mentioned, this could also be a common pattern in Vue. 👌